mockserver_test.go

```diff
@@ -307,22 +357,43 @@ func newSSHServerShell(conn net.Conn) (*sshServerShell, error) {
 			}
 		}(reqs)
 
+		c.mtx.Lock()
 		c.sshChannel = sChan
+		c.mtx.Unlock()
 		c.cond.Broadcast()
 	}()
 
 	return c, nil
 }
 
-// Read reads from the ssh channel.
-func (c *sshServerShell) Read(b []byte) (int, error) {
-	for c.sshChannel == nil && !c.closed {
+func (c *sshServerShell) channel() (ssh.Channel, bool) {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	return c.sshChannel, c.closed
+}
+
+func (c *sshServerShell) waitChannel() (ssh.Channel, error) {
+	ch, closed := c.channel()
+	for ch == nil {
+		if closed {
+			return nil, net.ErrClosed
+		}
 		c.cond.Wait()
+		ch, closed = c.channel()
```

Comment:
There should probably be another check whether `closed` is true. Theoretically (even if unlikely) there could be a `channel` which is already closed.

Another possibility is to return `nil` from `c.channel()` if closed.

Reply:
Nice catch, fixed.

Also triggered another race when testing this fix, so there's additional validation to ignore `connection reset by peer on close`